presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
 package com.facebook.presto.cost;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Double.NEGATIVE_INFINITY;
+import static java.lang.Double.NaN;
+import static java.lang.Double.POSITIVE_INFINITY;
+import static java.lang.Double.isNaN;
+import static java.lang.Double.max;
+import static java.lang.Double.min;
 import static java.util.stream.Stream.concat;
 
 public class PlanNodeStatsEstimateMath
 {
     private PlanNodeStatsEstimateMath()
     {}
 
-    @FunctionalInterface
-    private interface RangeSubtractionStrategy
+    public static PlanNodeStatsEstimate subtractStats(PlanNodeStatsEstimate superset, PlanNodeStatsEstimate subset)
     {
-        StatisticRange range(StatisticRange leftRange, StatisticRange rightRange);
+        if (superset.isOutputRowCountUnknown() || subset.isOutputRowCountUnknown()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        double supersetRowCount = superset.getOutputRowCount();
+        double subsetRowCount = subset.getOutputRowCount();
+        double outputRowCount = supersetRowCount - subsetRowCount;
+        verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
+
+        // everything will be filtered out after applying negation
+        if (outputRowCount == 0) {
+            return createZeroStats(superset);
+        }
+
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        result.setOutputRowCount(outputRowCount);
+
+        superset.getSymbolsWithKnownStatistics().forEach(symbol -> {
+            SymbolStatsEstimate supersetSymbolStats = superset.getSymbolStatistics(symbol);
+            SymbolStatsEstimate subsetSymbolStats = subset.getSymbolStatistics(symbol);
+
+            SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
+
+            // for simplicity keep the average row size the same as in the input
+            // in most cases the average row size doesn't change after applying filters
+            newSymbolStats.setAverageRowSize(supersetSymbolStats.getAverageRowSize());
+
+            // nullsCount
+            double supersetNullsCount = supersetSymbolStats.getNullsFraction() * supersetRowCount;
+            double subsetNullsCount = subsetSymbolStats.getNullsFraction() * subsetRowCount;
+            double newNullsCount = supersetNullsCount - subsetNullsCount;
+            verify(isNaN(newNullsCount) || newNullsCount >= 0, "newNullsCount must greater than or equal to zero: %s", newNullsCount);
+            newSymbolStats.setNullsFraction(min(newNullsCount, outputRowCount) / outputRowCount);
+
+            // distinctValuesCount
+            double supersetDistinctValues = supersetSymbolStats.getDistinctValuesCount();
+            double subsetDistinctValues = subsetSymbolStats.getDistinctValuesCount();
+            double newDistinctValuesCount;
```

Comment:
this can be simplified, see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R75

Reply:
I'm trying to avoid division by zero. In the simplified version division by zero may happen.